packages/vincent-tools/vincent-tool-sdk/src/lib/get-onchain-policy-params.ts

```diff
@@ -0,0 +1,97 @@
+import { z } from "zod";
+import { ethers } from "ethers";
+
+import type { EthersAbiDecodedValue, Policy, VincentPolicyDef } from "./types";
+import { abiDecodePolicyParameters } from "./abi-decode-policy-params";
+import { formatZodError } from "./format-zod-error";
+
+interface AllOnChainPolicyParams {
+    isPermitted: boolean;
+    appId: ethers.BigNumber;
+    appVersion: ethers.BigNumber;
+    policies: Policy[];
+}
+
+export const getOnChainPolicyParams = async (args: {
+    yellowstoneRpcUrl: string,
+    vincentContractAddress: string,
+    appDelegateeAddress: string,
+    agentWalletPkpTokenId: string,
+    toolIpfsCid: string,
+    policyIpfsCid: string,
+    policyUserParamsSchema?: z.infer<VincentPolicyDef['userParamsSchema']>
+}): Promise<z.infer<VincentPolicyDef['userParamsSchema'] | undefined>> => {
+    try {
+        const allOnChainPolicyParams = await _getAllOnChainPolicyParams(args);
+
+        // We exit early here because !allOnChainPolicyParams.isPermitted means appDelegateeAddress
+        // is not permitted to execute toolIpfsCid for the Vincent App on behalf of the agentWalletPkpTokenId
+        // and no further processing is needed
+        if (!allOnChainPolicyParams.isPermitted) {
+            throw new Error(`App Delegatee: ${args.appDelegateeAddress} is not permitted to execute Vincent Tool: ${args.toolIpfsCid} for App ID: ${allOnChainPolicyParams.appId.toString()} App Version: ${allOnChainPolicyParams.appVersion.toString()} using Agent Wallet PKP Token ID: ${args.agentWalletPkpTokenId}`);
+        }
+
+        const onChainPolicyParams = allOnChainPolicyParams.policies.find(
```

Comment:
Now that we're running this per-policy, can we use the contract function you mentioned that just returns params for this one policy, rather than relying on a .find() to select it? We're getting a lot more data here than we need.

Reply:
There isn't a function to return a single policies' parameter, we only have the single read function that returns all the  of the registered policy params for all the tools for an App for a specific Agent Wallet